Integrate Frontend Video Summary Feature

Uncle3k opened the pull request:
Integrate Frontend Video Summary Feature

This pull request implements the frontend integration for the video summary feature, enabling users to generate a summary of the currently viewed YouTube video. It establishes seamless communication with the backend `/ai/generate-summary` endpoint and ensures a clean, user-friendly UI. Below are the key changes:

---

### **Files Changed:**

1. **`chatContainer.js`**
    - **Updated Function:** `createChatContainer`
        - Added a new "Generate Summary" button to the chatbot's header.
        - Integrated a click event listener to trigger the video summary feature by communicating with the backend.
    - **Added Functionality:**
        - Handles interactions to fetch video summaries.
        - Provides user feedback in cases of errors like missing transcripts or backend failures.
2. **`content.js`**
    - **Added Function:** `generateVideoSummary`
        - Extracts the `video_id` from the YouTube URL using the existing `extractVideoID` function.
        - Sends a properly formatted POST request to the backend `/ai/generate-summary` endpoint with `video_id`.
        - Handles responses and errors gracefully, updating the chatbot UI as needed.
3. **`styles.css`**
    - **Added Styling:**
        - Add a temporary book as the "Generate Summary" button.
        - Included hover effects and alignment updates for better user experience.
4. **`webpack.config.js`**
    - Ensured the new functionality is bundled into `content.bundle.js` during the Webpack build process.

---

### **Functions Added:**
1. **`generateVideoSummary` (in `content.js`):**
    - Extracts `video_id` from the video URL.
    - Communicates with the backend to retrieve the video summary.
    - Implements robust error handling for scenarios such as:
        - Missing or invalid `video_id`.
        - Backend issues like missing transcripts or failed summary generation.

---

### **Key Features:**

1. **New UI Element:**
    - Added a "Generate Summary" button in the chatbot header, opposite the menu button (☰).
    - Designed to fit the chatbot's UI theme with minimalistic styling and hover effects.
2. **Backend Integration:**
    - Sends the required `video_id` to the backend endpoint.
    - Handles responses and displays the summary or error messages in the chatbot interface.
3. **Error Handling:**
    - Notifies the user if the transcript is unavailable.
    - Handles scenarios like network failures or backend errors with clear user feedback.
4. **Code Improvements:**
    - Debug logs added for easier troubleshooting.
    - Clean and modular implementation, adhering to existing project structure and coding standards.

---

### **Testing:**

1. **Functionality:**
    - Tested with:
        - Valid video IDs (transcripts available).
        - Network and backend errors.
    - Verified that the chatbot provides accurate summaries or meaningful error messages.
2. **Compatibility:**
    - Confirmed the "Generate Summary" button is fully responsive and fits seamlessly into the chatbot UI.
3. **Build Verification:**
    - Rebuilt the project using Webpack to include new functionality in `content.bundle.js`.

AnasKhan0607 reviewed (changes requested):
Looks good but less change it and add it to the container Ruilin made here: 
https://github.com/Learning-Mode-AI/Learning-Mode-AI/pull/10

Uncle3k commented:
Move Video Summary Feature to Container2


This pull request implements changes to integrate the video summary feature into `Container2` (`features-panel`) while keeping the `ChatContainer` functionality intact. The changes improve the modularity of features and enhance the user interface for accessing video summaries via the dropdown in `Container2`.

---

**Changes:**

1. **`content.js`:**
    
    - Added logic to manage `Container2` alongside `ChatContainer`.
    - Updated `generateVideoSummary` function to display summaries in `Container2`'s `#summary-holder`.
    - Ensured `ChatContainer` retains its existing functionalities (e.g., AI Q&A).
2. **`chatContainer.js`:**
    
    - Removed the summary generation button to avoid redundancy.
    - Focused the container exclusively on chat-related functionalities.
3. **`container2.js`:**
    
    - Enhanced the dropdown to include a "Short Summary" option.
    - Integrated `generateVideoSummary` to fetch and display the video summary within the `#summary-holder`.
4. **`styles.css`:**
    
    - Verified existing styles for `Container2` (`#features-panel`) support the new video summary feature without additional changes.

---

**Testing Instructions:**

1. Toggle the learning mode to activate `Container2` and `ChatContainer`.
2. In `Container2`, use the dropdown to select "Short Summary."
3. Verify the video summary is displayed correctly in the `#summary-holder` of `Container2`.
4. Test the `ChatContainer` for chat-related functionalities to ensure no regressions.

---

**Impact:**

- **Improved User Experience**: Separates video summary and chat functionalities for clearer usage.
- **Code Maintainability**: Modularizes responsibilities for better feature management.

AnasKhan0607 reviewed (changes requested):
let's test for overflow text for longer summaries

Uncle3k commented:
This PR adds Redis caching to prevent redundant API calls to the AI service for video summaries. The `VideoSummaryHandler` now checks if a summary exists in Redis before making an API request to the AI service. If a cached summary exists, it is returned directly; otherwise, a new summary is requested and then cached for future use.

**Key Changes:**

- **New Redis Key:** `"summary:{videoID}"` (stores video summary as a string with a 24-hour TTL)
- Updated `videoSummaryHandler.go` to check Redis for cached summaries.
- Added `StoreVideoSummaryInRedis` and `GetVideoSummaryFromRedis` in `redis.go`.

**Data Flow:**

1. **Frontend Check:** The frontend checks local storage for a video summary.
2. **Backend Check:** If no summary exists locally, the frontend sends a request to the backend.
3. **Redis Cache Check:** The backend checks Redis for `"summary:{videoID}"`.
4. **Cache Hit:** If the summary exists, it is returned directly to the client.
5. **Cache Miss:** If no summary is found, the backend calls the AI service to generate a summary.
6. **AI Service Handling:** The AI service checks its own Redis cache before generating the summary.
7. **Caching:** If the summary is generated, it is stored in both the AI service and the backend's Redis caches for future use.

Uncle3k commented:
### 🚀 **Enhanced Feature Handling, Interest Tracking & Styling Consistency**

**Summary:**  
This PR introduces substantial enhancements to the backend and frontend, including improved feature management, user interaction tracking, and UI styling fixes across multiple components.

---

### 🎯 **Key Changes & Improvements:**

#### ✅ **Backend Enhancements:**

- **Redis Integration (`redis.go`):**
    - Added support for storing and retrieving **video summaries** and **user interest clicks** with unique keys for each feature.
    - Introduced **data sanitization** for consistent Redis keys by replacing spaces with underscores.
- **Router Expansion (`router.go`):**
    - Added new route for handling interest clicks:
        - `POST /api/show-interest`
        - `GET /api/interest-count/{feature}`
- **Interest Handling (`interestHandler.go`):**
    - Implemented the **`ShowInterestHandler`** for recording user interest clicks and validating requests.
    - Added the **`GetInterestCountHandler`** for retrieving interest counts per feature from Redis.

---

#### ✅ **Frontend Improvements:**

- **Feature Styling Refinements (`container2.js`):**
    - Introduced a **separation of styling** between `"Coming Soon"` and **functional features** using the `.coming-soon` and `.scrollable` classes.
    - `"Short Summary*"` and other functional features can now display a **scroll bar** while `"Coming Soon"` features remain scroll-free.
    - Ensured **left alignment** for the summary content.
- **Dropdown Management Fix:**
    - Corrected a **z-index** bug where the dropdown was hidden behind content.

---

### 🎨 **Styling Fixes (`styles.css`):**

- **Consistent Custom Scrollbar:**
    - Added a **thin scrollbar** with the color `#ECB0B0`.
    - **Rounded edges** and **hidden up/down arrows** were applied for a cleaner look.
- **Feature-Specific Styling Isolation:**
    - `.scrollable-feature` and `.coming-soon` classes ensure unique styling per feature type.

---

### 📦 **Impact:**

- **Improved User Experience:** Clearer visual distinctions between `"Coming Soon"` and **functional** features.
- **Data Consistency:** Properly sanitized Redis keys for accurate data storage.
- **Bug Fixes:** Fixed disappearing scrollbar and dropdown overlap issues

AnasKhan0607 commented:
Ill Fix some of the small things and conflicts and merge tomorow

nnourr commented on extention/js/content.js:
```diff
@@ -104,40 +113,97 @@ export function askAIQuestion(videoUrl, question) {
+            summaryHolder.style.display = 'block';
+            loadingIndicator.style.display = 'none';
         }
+        onSuccess && onSuccess(marked(storedSummary));
```
I see! we probably forgot to include the summary when calling onSuccess, and that's why it didn't work. but, you already create the `marked` version of the summary in the onSuccess function, so calling `marked` here too might be redundant.

nnourr replied:
actually, I just realized that you're doing everything here (setting the display to block and allowing scroll) in the onSuccess callback in container2.js. This duplicated logic could cause bugs in the future. Can you check if it still works without these lines:
```go
        if (summaryHolder) {
            summaryHolder.innerHTML = marked(storedSummary);
            summaryHolder.style.display = 'block';
            loadingIndicator.style.display = 'none';
        }
```

nnourr commented on extention/js/content.js:
```diff
@@ -104,40 +113,97 @@ export function askAIQuestion(videoUrl, question) {
+            if (data.summary) {
+                localStorage.setItem(`summary_${videoId}`, data.summary);
+                summaryHolder.innerHTML = marked(data.summary); // Render using 'marked'
+                summaryHolder.style.display = 'block';
```
same duplicated logic here

nnourr commented on extention/js/content.js:
```diff
@@ -104,40 +113,97 @@ export function askAIQuestion(videoUrl, question) {
+        .catch((error) => {
+            console.error('Error fetching video summary:', error.message);  // Use 'error.message' for clarity
+            if (summaryHolder) {
+                summaryHolder.innerText = 'Error fetching video summary.';
```
the same duplicated logic also applies to the error states

nnourr commented:
> This is a very big PR so its hard to review. Specially the frontend code. Next time lets try if possible to make them smaller and split things where we can.

I agree, this would've been a great opportunity to create tickets for things you've noticed that aren't directly related to the ticket (i.e. changing `AI Generated` to `Reinforce your learning`)
